Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,497 @@
+// Blender uses hardcoded string identifiers to figure out what the sockets do:
+const auto baseColorIdentifier = "Base Color";
+const auto colorIdentifier = "Color";
+const auto roughnessIdentifier = "Roughness";
+const auto metallicIdentifier = "Metallic";
+const auto normalIdentifier = "Normal";
+const auto normalStrengthIdentifier = "Strength";
+const auto surfaceIdentifier = "Surface";
+const auto emissionIdentifier = "Emission";
+const auto emissionStrengthIdentifier = "Emission Strength";
+
+const auto shaderIdentifier = "Shader";
+
+const auto displacementIdentifier = "Displacement";
+const auto heightIdentifier = "Height";
+const auto scaleIdentifier = "Scale";
+
+// Removes reroute nodes and returns the actual input node upstream.
+bNode* removeReroutes(bNode* node, const Material* mat) {
+	if (!node || node->type != NODE_REROUTE) {
+		return node;
+	}
+
+	auto tree = mat->nodetree;
+
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode == node) {
+			return removeReroutes(link->fromnode, mat);
+		}
+	}
+
+	return nullptr;
+}
+
+bNode* getNodeConnectedToSocket(bNodeSocket* socket) {
+	if (socket->link) {
+		return socket->link->fromnode;
+	}
+
+	return nullptr;
+}
+
+// For mix shaders, we pass through the first input that has a connection.
+bNode* handleMixShader(const Material* mat, bNode* bsdf) {
+	if (!bsdf || bsdf->type != SH_NODE_MIX_SHADER)
+		return bsdf;
+
+	for (auto inputSocket : list_range((bNodeSocket*)bsdf->inputs.first)) {
+		if (STREQ(inputSocket->name, shaderIdentifier)) {
+			bNode* connectedBSDF = removeReroutes(getNodeConnectedToSocket(inputSocket), mat);
+			if (connectedBSDF)
+			{
+				return connectedBSDF;
+			}
+		}
+	}
+
+	return bsdf;
+}
+
+bool getBSDFAndOutput(const Material* mat, bNode*& bsdf, bNode*& output) {
+	// Find BSDF connected to an output node:
+	auto tree = mat->nodetree;
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode &&
+			link->tonode->type == SH_NODE_OUTPUT_MATERIAL &&
+			STREQ(link->tosock->identifier, surfaceIdentifier)) {
+			bsdf = removeReroutes(link->fromnode, mat);
+
+			bsdf = handleMixShader(mat, bsdf);
+
+			output = link->tonode;
+			return bsdf && output;
+		}
+	}
+
+	return false;
+}
+
+bNodeSocket* getInputSocket(bNode* node, const char* socketName) {
+	for (auto inputSocket : list_range((bNodeSocket*)node->inputs.first)) {
+		if (STREQ(inputSocket->name, socketName)) {
+			return inputSocket;
+		}
+	}
+
+	return nullptr;
+}
+
+int msblenMaterialsExportHelper::exportTexture(const std::string& path, ms::TextureType type) const
+{
+	return m_texture_manager->addFile(path, type);
+}
+
+void msblenMaterialsExportHelper::exportPackedImages(ms::TextureType& textureType,
+	std::function<void(int textureId)> setTextureHandler,
+	Image* img) const
+{
+	for (auto imagePackedFile : list_range((ImagePackedFile*)img->packedfiles.first)) {
+		std::string name = img->id.name + 2; // +2 because of IM prefix!
```

Comment:
What sort of black magic is this? 

Does this create a string in the variable `name` that is equal to `img->id.name` but two empty characters longer?

Reply:
Or is this doing some pointer arithmetic from the start of the `img->id.name` character array?

Reply:
I have used the same pattern before, its pointer arithmetic as you mentioned.

Reply:
Recommend taking another stab at this as this is a bit dangerous as it is currently written, right now it's assuming the string is at least 2 characters long when there's no guarantee that that's always the case, so putting a length check in would be prudent to avoid a possible segfault in that scenario.

Recommend also avoiding the pointer arithmetic simply because its kinda gross and I was raised to abhor it, if you choose to do that std::string_view is your friend as it allows you to use a c-style string like a std::string (with an iterator) and it has the added bonus of not copying anywhere, which in this scenario is a micro-optimization so not too important. You can read more about that here https://en.cppreference.com/w/cpp/string/basic_string_view

Reply:
Yes, blender has 2 character ID prefixes in front of every name of an object that need to be skipped.
I've refactored it to use a string_view now.